### PR TITLE
Optimize metadata

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/KAC-39

Changes:
- Config metadata remote load/save have been moved from `mapper` to `remote` package.
- Config metadata are loaded in parallel with configs.
- Config metadata are saved only if needed.

### CODE REVIEW

File: internal/pkg/api/storageapi/config.go [hunk 1]
```diff
@@ -123,8 +123,13 @@ func (c ConfigMetadataResponse) MetadataMap(branchId model.BranchId) map[model.C
 	return result
 }
 
-// UpdateConfigMetadataRequest https://keboola.docs.apiary.io/#reference/metadata/components-configurations-metadata/create-or-update
-func (a *Api) UpdateConfigMetadataRequest(config *model.Config) *client.Request {
+// AppendConfigMetadataRequest https://keboola.docs.apiary.io/#reference/metadata/components-configurations-metadata/create-or-update
```
Comment (webrouse) [inline internal/pkg/api/storageapi/config.go:126]: Premenoval som to na `Append`, kedze nove hodnoty sa pridaju/prepisu, stare ostanu.

File: internal/pkg/mapper/configmetadata/remote.go [hunk 2]
```diff
@@ -1,44 +0,0 @@
-package configmetadata
-
-import (
-	"github.com/keboola/keboola-as-code/internal/pkg/model"
-)
-
-func (m *configMetadataMapper) AfterRemoteOperation(changes *model.RemoteChanges) error {
-	// Process saved objects
```
Comment (webrouse) [inline internal/pkg/mapper/configmetadata/remote.go]: Remote save presunute do `remote` pkg.

File: internal/pkg/state/remote/manager.go [hunk 3]
```diff
@@ -283,22 +283,53 @@ func (u *UnitOfWork) createOrUpdate(objectState model.ObjectState, object model.
...
 
-	// Update
-	return u.update(objectState, object, recipe, changedFields)
+	// Send
+	createOrUpdateReq.Send()
+	return nil
 }
 
-func (u *UnitOfWork) create(objectState model.ObjectState, object model.Object, recipe *model.RemoteSaveRecipe) error {
+func (u *UnitOfWork) createRequest(objectState model.ObjectState, object model.Object, recipe *model.RemoteSaveRecipe) (*client.Request, error) {
```
Comment (webrouse) [inline internal/pkg/state/remote/manager.go:326]: Neposila request, iba ho vytvara. Podobne `updateRequest` ... kedze potrebujeme napojit ulozenie metadat.

File: internal/pkg/mapper/configmetadata/remote_test.go [hunk 4]
```diff
@@ -1,94 +0,0 @@
...
-	mockedState.Mapper().AddMapper(configmetadata.NewMapper(mockedState, d))
-	return mockedState, httpTransport
-}
-
-func TestConfigMetadataOnRemoteChangeSaved(t *testing.T) {
-	t.Parallel()
-	mockedState, httpTransport := InitState(t)
-
-	content := orderedmap.New()
-	json.MustDecodeString("{}", content)
```
Comment (webrouse) [inline internal/pkg/mapper/configmetadata/remote_test.go:120]: Presunuty test do `remote` pkg.

File: internal/pkg/state/remote/manager_test.go [hunk 5]
```diff
@@ -281,6 +281,152 @@ func TestLoadConfigMetadata(t *testing.T) {
 	assert.Equal(t, map[string]string{}, projectState.MustGet(config2Key).(*model.ConfigState).Remote.Metadata)
 }
 
+func TestSaveConfigMetadata_Create(t *testing.T) {
+	t.Parallel()
+	uow, httpTransport, _ := newTestRemoteUOW(t)
+
```
Comment (webrouse) [inline internal/pkg/state/remote/manager_test.go:287]: Presunute/doplnene testy.
Request na append metadata sa odosiela, iba ked nie su metadata prazdne + nieco sa zmenilo.

**Review by webrouse — COMMENTED** [review]
Pozri prosim po commitoch.

Tuto pracu sme minule odlozili. 
Metadata set request sa posielal vzdy. 
Metadata get request nebezal paralelne s nacitanim konfiguracii.

File: internal/pkg/http/client/pool_test.go [hunk 6]
```diff
@@ -208,6 +208,48 @@ func TestErrorInSubRequest(t *testing.T) {
...
+	httpTransport.RegisterResponder("GET", `https://example.com`, httpmock.NewStringResponder(200, `test`))
+	httpTransport.RegisterResponder("GET", `https://example.com/error`, httpmock.NewErrorResponder(errors.New("network error")))
+
+	// Create pool and requests
+	pool := client.NewPool(logger)
+	request1 := pool.Request(client.NewRequest(resty.MethodGet, "https://example.com"))
+	request2 := pool.Request(client.NewRequest(resty.MethodGet, "https://example.com/error"))
+
+	// Request 1 wait for request 2
+	request1.WaitFor(request2)
```
[the remark this answers is not in the document]
Comment (JakubMatejka) [reply internal/pkg/http/client/pool_test.go:223]: Aha, super. To mi předtím nebylo jasný jak to v Go vlastně funguje.

**Review by JakubMatejka — APPROVED** [review]
👍

File: internal/pkg/http/client/pool_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (webrouse) [reply internal/pkg/http/client/pool_test.go:223]: To je skor nasa "custom" vec. V Go na to nie je ziadna dobra kniznica ani postup.
Su `channels` a nad tym si to mozes postavit.
Len my to potrebujeme na strasne vela miestach, preto su `channels` zabalene do `pool` pkg.

https://medium.com/dev-bits/making-concurrent-http-requests-in-go-programming-language-823b51bb1dc2

**Pull request merged**